proposals/infra/INF-0002-root-signature-in-clang.md

```diff
@@ -168,41 +168,64 @@ to ensure that our solution doesn't unnecessarily tie the non-HLSL parts to it.
            bReg (',' 'space' '=' NUMBER)?
            (',' 'visibility' '=' SHADER_VISIBILITY)? ')'
 
+    ROOT_DESCRIPTOR_FLAGS : 0 | 'DATA_STATIC' |
+                            'DATA_STATIC_WHILE_SET_AT_EXECUTE' |
+                            'DATA_VOLATILE'
+
     RootCBV : 'CBV' '(' bReg (',' 'space' '=' NUMBER)?
           (',' 'visibility' '=' SHADER_VISIBILITY)?
-          (',' 'flags' '=' DATA_FLAGS)? ')'
+          (',' 'flags' '=' ROOT_DESCRIPTOR_FLAGS)? ')'
 
     RootSRV : 'SRV' '(' tReg (',' 'space' '=' NUMBER)?
           (',' 'visibility' '=' SHADER_VISIBILITY)?
-          (',' 'flags' '=' DATA_FLAGS)? ')'
+          (',' 'flags' '=' ROOT_DESCRIPTOR_FLAGS)? ')'
 
     RootUAV : 'UAV' '(' uReg (',' 'space' '=' NUMBER)?
           (',' 'visibility' '=' SHADER_VISIBILITY)?
-          (',' 'flags' '=' DATA_FLAGS)? ')'
+          (',' 'flags' '=' ROOT_DESCRIPTOR_FLAGS)? ')'
 
     DescriptorTable : 'DescriptorTable' '(' (DTClause(|DTClause)?)?
           (',' 'visibility' '=' SHADER_VISIBILITY)? ')'
 
     DTClause : CBV | SRV | UAV | Sampler
 
+    DESCRIPTOR_RANGE_FLAGS :
+        0 |
+        'DESCRIPTORS_VOLATILE' |
+        'DATA_VOLATILE' |
+        'DATA_STATIC' |
+        'DATA_STATIC_WHILE_SET_AT_EXECUTE' |
+        'DESCRIPTORS_VOLATILE' '|' 'DATA_VOLATILE' |
+        'DESCRIPTORS_VOLATILE' '|' 'DATA_STATIC_WHILE_SET_AT_EXECUTE' |
+        'DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS' |
+        'DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS' '|' 'DATA_VOLATILE' |
+        'DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS' '|' 'DATA_STATIC' |
+        'DESCRIPTORS_STATIC_KEEPING_BUFFER_BOUNDS_CHECKS' '|' 'DATA_STATIC_WHILE_SET_AT_EXECUTE'
```

Comment:
This doesn't look like a grammar to me.  I think you're trying to list possible valid combinations of flags, with one valid combination per line?  But I don't think you're saying that we'd reject `DATA_VOLATILE | DESCRIPTORS_VOLATILE`?  This seems like a mix of formal and informal notation here.

I think you should keep it simple - have the grammar able to capture all the values of the `D3D12_DESCRIPTOR_RANGE_FLAG` enum, and then treat validating the flag combinations as part of semantic analysis.  The API doesn't have a distinction between sampler range flags and descriptor range flags, I don't think that the grammar should either.

Reply:
Updated.